HoRNDIS.cpp

```diff
@@ -773,66 +723,83 @@ bool HoRNDIS::configureInterface(IONetworkInterface *netif) {
 IOReturn HoRNDIS::getPacketFilters(const OSSymbol *group, UInt32 *filters) const {
 	IOReturn	rtn = kIOReturnSuccess;
 	
-	if (group == gIOEthernetWakeOnLANFilterGroup)
+	if (group == gIOEthernetWakeOnLANFilterGroup) {
 		*filters = 0;
-	else if (group == gIONetworkFilterGroup)
-		*filters = kIOPacketFilterUnicast | kIOPacketFilterBroadcast | kIOPacketFilterMulticast | kIOPacketFilterPromiscuous;
-	else
+	} else if (group == gIONetworkFilterGroup) {
+		// Not yet supporting Multicast: need more code for proper support.
+		// kIOPacketFilterMulticast
+		*filters = kIOPacketFilterUnicast | kIOPacketFilterBroadcast
+			| kIOPacketFilterPromiscuous;
+	} else {
 		rtn = super::getPacketFilters(group, filters);
+	}
 
 	return rtn;
 }
 
 IOReturn HoRNDIS::getMaxPacketSize(UInt32 * maxSize) const {
-	*maxSize = mtu;
+	IOReturn rc = super::getMaxPacketSize(maxSize);
+	if (rc != kIOReturnSuccess) {
+		return rc;
+	}
+	// The max packet size is limited by RNDIS max transfer size:
+	*maxSize = min(*maxSize, maxOutTransferSize - sizeof(rndis_data_hdr));
+	LOG(V_DEBUG, "returning %d", *maxSize);
 	return kIOReturnSuccess;
 }
 
 IOReturn HoRNDIS::selectMedium(const IONetworkMedium *medium) {
+	LOG(V_DEBUG, ">");
 	setSelectedMedium(medium);
 	
 	return kIOReturnSuccess;
 }
 
 IOReturn HoRNDIS::getHardwareAddress(IOEthernetAddress *ea) {
+	LOG(V_DEBUG, ">");
 	UInt32	  i;
 	void *buf;
 	unsigned char *bp;
 	int rlen = -1;
 	int rv;
 	
-	buf = IOMalloc(RNDIS_CMD_BUF_SZ);
-	if (!buf)
+	buf = IOMallocAligned(RNDIS_CMD_BUF_SZ, sizeof(void *));
+	if (!buf) {
 		return kIOReturnNoMemory;
-	
+	}
+
+	// WARNING: Android devices may randomly-generate RNDIS MAC address.
+	// The function may return different results for the same device.
+
 	rv = rndisQuery(buf, OID_802_3_PERMANENT_ADDRESS, 48, (void **) &bp, &rlen);
 	if (rv < 0) {
 		LOG(V_ERROR, "getHardwareAddress OID failed?");
-		IOFree(buf, RNDIS_CMD_BUF_SZ);
+		IOFreeAligned(buf, RNDIS_CMD_BUF_SZ);
 		return kIOReturnIOError;
 	}
 	LOG(V_DEBUG, "MAC Address %02x:%02x:%02x:%02x:%02x:%02x -- rlen %d",
 	      bp[0], bp[1], bp[2], bp[3], bp[4], bp[5],
 	      rlen);
 	
-	for (i=0; i<6; i++)
+	for (i=0; i<6; i++) {
 		ea->bytes[i] = bp[i];
+	}
 	
-	IOFree(buf, RNDIS_CMD_BUF_SZ);
+	IOFreeAligned(buf, RNDIS_CMD_BUF_SZ);
 	return kIOReturnSuccess;
 }
 
 IOReturn HoRNDIS::setPromiscuousMode(bool active) {
-	(void) active;
-	
-	/* XXX This actually needs to get passed down to support 'real' RNDIS devices, but it will work okay for Android devices. */
-	
+	// XXX This actually needs to get passed down to support 'real'
+	//  RNDIS devices, but it will work okay for Android devices.
 	return kIOReturnSuccess;
 }
 
+// TODO(mikhailai): Test and possibly fix suspend and resume.
+
+/*
```

Comment:
`#if 0`?  (or remove, if unnecessary?)

Reply:
Ok, "#if 0" I was planning to work on checking suspend/resume after releasing the current code.